src/tools/rbd_mirror/image_sync/ImageCopyRequest.cc

```diff
@@ -146,7 +148,20 @@ void ImageCopyRequest<I>::send_object_copies() {
       }
     }
     complete = (m_current_ops == 0);
+
+    if (!complete) {
+      m_update_sync_ctx = new FunctionContext([this](int r) {
+          this->send_update_sync_point();
+      });
+    }
+  }
+
+  if (m_update_sync_ctx) {
+    Mutex::Locker l(*m_timer_lock);
+    m_timer->add_event_after(m_update_sync_point_interval,
```

Comment:
Well, if ```complete``` was false on line 150 and when you drop the lock ```handle_object_copy``` invokes and discovers it should complete. At this point, ```m_update_sync_ctx``` is not null, ```handle_object_copy``` cancels the event (which hasn't been scheduled yet), sets it to null (future revision), and continues.  Therefore, you would need to grab the timer lock before testing m_update_sync_ctx for null  on like 159.